gpytorch/models/model_list.py

```diff
@@ -91,8 +98,26 @@ def train_targets(self):
 
 
 def _get_tensor_args(*args):
+    arg_list = []
     for arg in args:
         if torch.is_tensor(arg):
-            yield (arg,)
+            arg_list.append((arg,))
         else:
-            yield arg
+            arg_list.append(arg)
+    return arg_list
+
+
+def _length_safe_zip(*args):
+    """Python's `zip(...)` with checks to ensure the arguments have
+    the same number of elements.
+
+    NOTE: This does not work with generators.
```

Comment:
An alternative would be to convert the elements that do not implement `__len__` into a list automatically. But I guess that could be problematic if they are long / take up a bunch of memory that we don't want to allocate.

Reply:
Yeah, I thought about that. For generators specifically, we would have to update the `args` directly, since once you call `list(generator)`, the generator is consumed and cannot be reused. The issue with that is `args` is a tuple, so we would have to first make that into a list, then iterate over its elements and update them. Eliminating the generator args seemed easier to me for this particular use case. If there are other possible use cases for this across the codebase, then it might be worth it though.